src/main/scala/com/cloudera/sparkts/ARIMA.scala

```diff
@@ -0,0 +1,586 @@
+        j += 1
+      }
+
+      // moving average terms
+      j = 0
+      while (j < q) {
+        yHat(i) = yHat(i) + maTerms(j) * coefficients(intercept + p + j)
+        dEdTheta(0, intercept + p + j) -= maTerms(j)
+        j += 1
+      }
+
+      error =  yVec(i) - yHat(i)
+      sigma2 += math.pow(error, 2) / n
+      updateMAErrors(maTerms, error)
+      // update gradient
+      gradient :+= (dEdTheta(0, ::) * error).t
+      // shift back error derivatives to make room for next period
+      dEdTheta(1 to -1, ::) := dEdTheta(0 to -2, ::)
+      // reset latest derivatives to 0
+      dEdTheta(0, ::) := 0.0
+      i += 1
+    }
+
+    gradient := gradient :/ -sigma2
+    gradient.toArray
+  }
+
+  /**
+   * Updates the error vector in place with a new (more recent) error
+   * The newest error is placed in position 0, while older errors "fall off the end"
+   * @param errs array of errors of length q in ARIMA(p, d, q), holds errors for t-1 through t-q
+   * @param newError the error at time t
+   * @return a modified array with the latest error placed into index 0
+   */
+  def updateMAErrors(errs: Array[Double], newError: Double): Unit= {
+    val n = errs.length
+    var i = 0
+    while (i < n - 1) {
+      errs(i + 1) = errs(i)
+      i += 1
+    }
+    if (n > 0) {
+      errs(0) = newError
+    }
+  }
+
+  /**
+   * Perform operations with the AR and MA terms, based on the time series `ts` and the errors
+   * based off of `goldStandard` or `errors`, combined with elements from the series  `dest`.
+   * Weights for terms are taken from the current model configuration.
+   * So for example: iterateARMA(series1, series_of_zeros,  _ + _ , goldStandard = series1,
+   * initErrors = null)
+   * calculates the 1-step ahead forecasts for series1 assuming current coefficients, and initial
+   * MA errors of 0.
+   * @param ts Time series to use for AR terms
+   * @param dest Time series holding initial values at each index
+   * @param op Operation to perform between values in dest, and various combinations of ts, errors
+   *           and intercept terms
+   * @param goldStandard The time series to which to compare 1-step ahead forecasts to obtain
+   *                     moving average errors. Default set to null. In which case, we check if
+   *                     errors are directly provided. Either goldStandard or errors can be null,
+   *                     but not both
+   * @param errors The time series of errors to be used as error at time i, which is then used
+   *               for moving average terms in future indices. Either goldStandard or errors can
+   *               be null, but not both
+   * @param initMATerms Initialization for first q terms of moving average. If none provided (i.e.
+   *                    remains null, as per default), then initializes to all zeros
+   * @return the time series resulting from the interaction of the parameters with the model's
+   *         coefficients.
+   */
+  //private
```

Comment:
woopsie. Flat out mistake. I had been testing something quickly in the console, so commented out private for that run. Completely forgot to uncomment. My bad